### PR TITLE
Add done callback pattern support to fuzzer

### PR DESCRIPTION
**Pull request by bertschneider**
Allow a fuzz target to request a done callback as second parameter. If
used, the fuzz target has to invoke the done callback to report its
finished execution and request the next input.

### CODE REVIEW

File: packages/fuzzer/start_fuzzing_async.cpp [hunk 1]
```diff
@@ -104,6 +105,63 @@ void CallJsFuzzCallback(Napi::Env env, Napi::Function jsFuzzCallback,
...
+                    std::make_exception_ptr(JSException()));
+              } else {
+                data->promise->set_value(nullptr);
+              }
+            });
+        auto result = jsFuzzCallback.Call({buffer, done});
+        if (result.IsPromise()) {
+          // If the fuzz target received a done callback, but also returned a
+          // promise, the callback could already have been called. In that case
+          // is_done_callback is already set. If is_resolved is also set, the
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply packages/fuzzer/start_fuzzing_async.cpp]: Done.

File: packages/fuzzer/start_fuzzing_async.cpp [hunk 2]
```diff
@@ -104,6 +105,63 @@ void CallJsFuzzCallback(Napi::Env env, Napi::Function jsFuzzCallback,
...
+                                .Get("length")
+                                .As<Napi::Number>()
+                                .Int32Value();
+      // In case more than one parameter is expected, the second one is
+      // considered to be a done callback to indicate finished execution.
+      if (parameterCount > 1) {
+        context->is_done_called = false;
+        auto done =
+            Napi::Function::New<>(env, [=](const Napi::CallbackInfo &info) {
+              // If the done callback based fuzz target also returned a promise
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply packages/fuzzer/start_fuzzing_async.cpp]: Done.

File: examples/jest_integration/integration.fuzz.js [hunk 3]
```diff
@@ -20,7 +20,15 @@
 const target = require("./target");
 
 describe("My describe", () => {
-	it.fuzz("My regression tests", async (data) => {
+	it.fuzz("My fuzz test", (data) => {
 		target.fuzzMe(data);
 	});
+
+	it.fuzz("My callback fuzz test", (data, done) => {
+		target.callbackFuzzMe(data, done);
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply examples/jest_integration/integration.fuzz.js:27]: Added timeout handling.

**Comment by bertschneider** [comment]
This PR got bigger than expected, sorry in advance :grimacing:

File: packages/jest-runner/fuzz.ts [hunk 4]
```diff
@@ -50,69 +59,131 @@ export const fuzz: FuzzTest = (name, fn) => {
 	// points to the element containing the fuzz function.
 	const testStatePath = currentTestStatePath(testName);
 
+	// The timeout setting is extracted from the test state and defaults to 5s.
+	// Setting timeouts on this level is necessary, as they apply to the
+	// individual inputs and not the whole run.
+	const timeout = currentTimeout();
```
Comment (oetr) [inline packages/jest-runner/fuzz.ts:65]: Do we want the ability to set the timeout like Jest does?
`it.fuzz("foo", (data) =>{}, 1000);`
It (almost) works for the hooks in the FUZZ-443-jest-hooks PR.
We "just" (!) need to forward it to each `runTestEntry()`.

**Review by oetr — APPROVED** [review]
A great addition to Jazzer.js, thanks :+1: 
My comments mostly point out the differences between Jest and our current state.
So, feel free to merge and open new tickets, or even address everything here. Whatever you see fit!

File: examples/jest_integration/integration.fuzz.js [hunk 5]
```diff
@@ -14,17 +14,51 @@
...
+		await target.asyncFuzzMe(data);
+	});
+
+	// In regression mode sync timeouts can not be detected, as the main event
+	// loop is blocked and registered timeout handlers can not fire.
+	// This is not only the case in regression test mode, but also during
+	// fuzzing runs. As the main event loop is blocked, no errors can be
+	// propagated to Jest. But the timeout set in libFuzzer will trigger a
+	// finding and shut down the whole process with exit code 70.
+	it.skip.fuzz("Sync timeout", () => {
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply examples/jest_integration/integration.fuzz.js:40]: Probably there is not much we can do here, at least not in a clean way. 

I added checks for multiple calls to the `done` callback which resolve the initial promise, so that Jest could finish the tests. In the synchronous case the event loop is still blocked by the endless loop and so the resolved promise can not be processed, hence, everything is stuck. In the async case we can not notify the fuzzer thread about the second invocation, so it keeps on running until the configured timeout. I also added a few comments for the new checks, clarifying why the test can not be shut down cleanly. At least no OOM is generated anymore. 

I think this is quite a corner case and the environment does not support a clean solution. I would like to try it out this way and see if problems arise.

File: examples/jest_integration/integration.fuzz.js [hunk 6]
```diff
@@ -14,17 +14,51 @@
...
 describe("My describe", () => {
-	it.fuzz("My regression test", async (data) => {
+	it.fuzz("My fuzz test", (data) => {
 		target.fuzzMe(data);
 	});
 
-	it.fuzz("My other regression test", (data) => {
-		target.fuzzMe(data);
+	it.fuzz("My callback fuzz test", (data, done) => {
+		target.callbackFuzzMe(data, done);
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply examples/jest_integration/integration.fuzz.js:27]: Good point! We now use the same formatting as the normal Jest runner, hence also display the code location, which is a really nice addition.

File: examples/jest_integration/integration.fuzz.js [hunk 7]
```diff
@@ -14,17 +14,51 @@
  * limitations under the License.
  */
 
-/* eslint no-undef: 0 */
+/* eslint no-undef: 0, no-constant-condition: 0, @typescript-eslint/no-var-requires:0 */
 
-// eslint-disable-next-line @typescript-eslint/no-var-requires
 const target = require("./target");
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply examples/jest_integration/integration.fuzz.js:19]: done

File: packages/jest-runner/fuzz.ts (hunk 4 above)
Comment (bertschneider) [reply packages/jest-runner/fuzz.ts:66]: Let's pull this feature out into a dedicated PR.
